### ISSUE TITLE
Incorrect monitor resolution being reported by display-info crate

### ISSUE THREAD

**Issue by THEGOLDENPRO** [cloudy-org/roseate#46]
If you're on a laptop with a 1920x1080 display and it's scale factor is set to 125% like mine the incorrect width and height will be reported by display-info breaking the smart downsampling optimization in Roseate.

See https://github.com/nashaofu/display-info/issues/38

_**notes** from here on..._

## Current Solution (egui viewport) and ways to deal with changes.
1. Get egui viewport monitor size initially, if no viewport monitor size initially available default to 1920x1080.
2. *window is active, egui viewport monitor size should be available now*
3. if it's available, cache that (persistent cache) as one of the user's monitor resolutions.
4. *user moves window to 2nd display (with a higher resolution, e.g. 4k)*
5. *egui viewport monitor size updates*
6. if the size is bigger than the previously cached size (a bigger monitor), cache that as another monitor size

### Roseate is launched a 2nd time
1. *we now have the cached monitor sizes*
2. initially use the highest monitor size from cache (it's better to assume a higher monitor size than lower)
3. unless the user's config says otherwise (some sort of `monitor_size_assumption_priority = "h"` key, where `h` picks the highest resolution, `m` picks the middle one and `l` picks the lowest

If something like `override_monitor_size` is set then **all** of these checks are prevented and egui viewport monitor size is ignored.

### Dynamic Sampling
While egui viewport monitor size is changing the image is also dynamically sampling itself to be the best quality for that current monitor. This will be called "Dynamic Sampling via Monitor Size".

## Progress
- [x] Cache egui viewport monitor size.
- [x] Update cached monitor size when new monitor size found.
- [x] Use cached monitor size initially.
- [x] ~~Add some sort of `monitor_size_assumption_priority = "h"` key.~~ (skipped)
- [x] Add `override_monitor_size` key.

**Comment by r3tr0ananas** [comment]
egui has it's own *"display-info" (it only returns the size of the display the app is on + you need access to a `Context`)*.

```rust
let monitor_size = ctx.input(|i| {
    let viewport = i.viewport();

    viewport.monitor_size
});
```

**Comment by THEGOLDENPRO** [comment]
> egui has it's own _"display-info" (it only returns the size of the display the app is on + you need access to a `Context`)_.
> 
> ```rust
> let monitor_size = ctx.input(|i| {
>     let viewport = i.viewport();
> 
>     viewport.monitor_size
> });
> ```

WHAT! Did they add that recently or like in the last 2 major releases? That's literally all I need.

**Comment by r3tr0ananas** [comment]
> > egui has it's own _"display-info" (it only returns the size of the display the app is on + you need access to a `Context`)_.
> > ```rust
> > let monitor_size = ctx.input(|i| {
> >     let viewport = i.viewport();
> > 
> >     viewport.monitor_size
> > });
> > ```
> 
> WHAT! Did they add that recently or like in the last 2 major releases? That's literally all I need.

It existed since v0.24.0: https://docs.rs/egui/0.24.0/egui/struct.ViewportInfo.html

**Comment by THEGOLDENPRO** [comment]
Now that I think of it I feel like I've tried this before and it didn't work or something, but I'll certainly try it again.

**Comment by THEGOLDENPRO** [comment]
Yeah I wanna do a complete restructure of `Image` and `ImageLoader` before I pass ctx.

**Comment by THEGOLDENPRO** [comment]
Was just messing around with this and `viewport.monitor_size` updates in real-time as the window is moved between differently sized monitors. ~~🔥 fucking fire~~

**Comment by THEGOLDENPRO** [comment]
It takes display-info about 5ms on my system to get the resolution of my display, so that's also a 5ms delay before roseate can load an image. That's really fast so I don't have a problem with that but as I'm switching to egui's viewport for this and so I can have monitor size available before the window launches, I'll have to cache... now I wonder if I can read from that cached file faster than the 5ms it would take display-info to get display / monitor size. 🤔 *just something that I've never really timed / benchmarked and now I'm interested to see the outcome of*

**Comment by THEGOLDENPRO** [comment]
This (#49) will be ready to merge soon just a few things remaining below. The "monitor_size_assumption_priority" config feature will be skipped as I now think it's quite unnecessary.

- Add `override_monitor_size` config key.
- Improve error handling and reporting (there's many `TODO` comments on this, I'll get to it, ideas are currently floating in my head on how to improve all of this)

**Comment by THEGOLDENPRO** [comment]
I think this issue has revisited itself on my laptop again, I'll create a new issue the next time I decide to chase this bug.
